[#149144401] Allow `CREATE SCHEMA`

henrytk opened the pull request:
## What

https://www.pivotaltracker.com/story/show/149144401

We want to allow people to execute CREATE SCHEMA in their broker-managed postgresql instances.

* Additional GRANT to the _manager group to give ALL priviledges to the database.
* ~~Changes the owner of the default "public" schema to the rds_xxx_manager group to allow users to remove it if they so wish.~~ (_see second commit message_)
* Tests to ensure CREATE/DROP schema is possible with the generated credentials
* Tests to ensure tables can be created and used within other schemas.

## How to review

* Check the code
* Run the tests
* Deploy to dev and maybe test with binding an app. If you want to deploy the broker we have a [build of the bosh release](https://concourse.build.ci.cloudpipeline.digital/teams/main/pipelines/rds-broker-release/jobs/build-dev-release/builds/61), and a [branch on paas-cf](https://github.com/alphagov/paas-cf/tree/create-schema-permissions) which pins said release.

## Who can review

not @chrisfarms or @henrytk

Commit:
Revert ownership of public schema

We currently have functions and triggers (reassigned_owned) installed
into the "public" schema.

Making the user the owner of the public schema enables them to
drop and cascadingly delete everything within the schema including any
funcs/triggers installed by the broker to manage object ownership.

We decided that it is not wise at this time to give full priviledges to
the public schema.

henrytk committed:
Bugfix: do not get status code on nil response

The integration tests were sometimes panicking due to trying to get
the status code when a nil response was returned.

dcarley commented on sqlengine/postgres_engine.go:
```diff
@@ -173,7 +190,7 @@ func (d *PostgresEngine) MigrateLegacyAdminUsers(tx *sql.Tx, bindingID, dbname s
 		d.logger.Debug("revoke-permissions", lager.Data{"statement": revokeStatement})
 
-		if _, err := d.db.Exec(revokeStatement); err != nil {
+		if _, err := tx.Exec(revokeStatement); err != nil {
```
Does this change belong to #62?

If so, I think it's worth breaking out to a separate commit.

henrytk replied:
Yes, we'll split it out.

henrytk replied:
Done

dcarley commented on ci/blackbox/rdsbroker_test.go:
```diff
@@ -306,50 +305,77 @@ func openConnection(databaseURI string) (*sql.DB, error) {
 		return fmt.Errorf("Error inserting record: %s", err.Error())
 	}
 
+	switch serviceID {
```
From offline discussion: @henrytk and I talked about the risk of these tests not being run if we changed the name of the service ID in the catalog (which is actually a string so that we can identify it easier). We decided that `openConnection` should extract the `scheme` from the URL and return that for us to use, because it comes from Compose and won't be changed.

henrytk replied:
On closer inspection this isn't actually pulled from the catalog, it's [hard coded](https://github.com/alphagov/paas-rds-broker/blob/b9d587b9090a86224e500786f397932c2664b603/ci/blackbox/rdsbroker_test.go#L124). We should still change it though because pulling it from the URI scheme is much less brittle.

henrytk replied:
Done, but I haven't tested it in any way. It's very late on a Friday afternoon, I'll sanity check it on Monday.

dcarley commented on ci/blackbox/rdsbroker_test.go:
```diff
@@ -306,50 +305,77 @@ func openConnection(databaseURI string) (*sql.DB, error) {
 		return fmt.Errorf("Error DROPing table: %s", err.Error())
 	}
 
+	switch serviceID {
```
Same as comment above about switching on scheme.

henrytk committed:
[#149144401] Enable CREATE/DROP SCHEMA schema

We want to allow people to execute CREATE SCHEMA in their broker-managed postgresql instances.
This change adds:

* Additional GRANT to the _manager group to give ALL priviledges to the database.
* Changes the owner of the default "public" schema to the rds_xxx_manager group to allow users to remove it if they so wish.
* Unit/Integration tests to ensure CREATE/DROP schema is possible with the generated credentials
* Unit/Integration tests to ensure tables can be created and used within other schemas.

henrytk committed:
Bugfix: put SQL command in transaction

This was omitted when working on concurrency improvements[1].

[1] https://github.com/alphagov/paas-rds-broker/pull/62

henrytk committed:
Use connection string URI to determine database

This is much less brittle than hardcoding 'postgres'. If somebody
were to change that string the tests would be skipped silently.

dcarley commented on ci/blackbox/rdsbroker_test.go:
```diff
@@ -306,50 +305,85 @@ func openConnection(databaseURI string) (*sql.DB, error) {
 }
 
-func setupPermissionsTest(databaseURI string) error {
+func setupPermissionsTest(databaseURI, serviceID string) error {
```
Can we remove the `serviceID` argument now?

dcarley commented on ci/blackbox/rdsbroker_test.go:
```diff
@@ -306,50 +305,85 @@ func openConnection(databaseURI string) (*sql.DB, error) {
 }
 
-func permissionsTest(databaseURI string) error {
+func permissionsTest(databaseURI, serviceID string) error {
```
Can we remove the `serviceID` argument now?

henrytk replied:
Yes. I've removed it and fixed up the change.
